src/dex/bebop/bebop.ts

```diff
@@ -0,0 +1,744 @@
+    }
+
+    return [];
+  }
+
+  runInstruction(
+    instruction: RoutingInstruction,
+    amount: BigNumber,
+  ): BigNumber {
+    let accumulated = BigNumber(0);
+    let output = BigNumber(0);
+    let filled = false;
+    for (const level of instruction.book.bids) {
+      const [price, size] = level;
+      const amountToAccumulate = instruction.targetQuote
+        ? BigNumber(size).times(price)
+        : BigNumber(size);
+      const afterAccumulated = accumulated.plus(amountToAccumulate);
+      if (afterAccumulated.lt(amount)) {
+        accumulated = accumulated.plus(amountToAccumulate);
+        const amountToAddToOutput = instruction.targetQuote
+          ? BigNumber(size)
+          : BigNumber(size).times(price);
+        output = output.plus(amountToAddToOutput);
+        if (accumulated.eq(amount)) {
+          filled = true;
+          break;
+        }
+      } else {
+        const remaining = amount.minus(accumulated);
+        output = output.plus(
+          instruction.targetQuote
+            ? remaining.div(price)
+            : remaining.times(price),
+        );
+        filled = true;
+        break;
+      }
+    }
+    if (filled) {
+      return output;
+    } else {
+      return BigNumber(0);
+    }
+  }
+
+  calculateOutput(
+    instructions: RoutingInstruction[],
+    srcToken: Token,
+    destToken: Token,
+    amounts: bigint[],
+    side: SwapSide,
+  ): bigint[] {
+    const outputs = [];
+    const inputDecimals =
+      side == SwapSide.SELL ? srcToken.decimals : destToken.decimals;
+    const outputDecimals =
+      side == SwapSide.SELL ? destToken.decimals : srcToken.decimals;
+
+    for (const amount of amounts) {
+      if (amount == 0n) {
+        outputs.push(0n);
+        continue;
+      }
+      const amountDecimals = BigNumber(amount.toString()).div(
+        getBigNumberPow(inputDecimals),
+      );
+      let output: BigNumber = BigNumber(0);
+      for (const instruction of instructions) {
+        output = this.runInstruction(
+          instruction,
+          output.gt(0) ? output : amountDecimals,
+        );
+        if (output.eq(0)) {
+          break;
+        }
+      }
+      if (output.gt(0)) {
+        outputs.push(
+          BigInt(output.times(getBigNumberPow(outputDecimals)).toFixed(0)),
+        );
+      } else {
+        outputs.push(0n);
+      }
+    }
+    return outputs;
+  }
+
+  // Returns pool prices for amounts.
+  // If limitPools is defined only pools in limitPools
+  // should be used. If limitPools is undefined then
+  // any pools can be used.
+  async getPricesVolume(
+    srcToken: Token,
+    destToken: Token,
+    amounts: bigint[],
+    side: SwapSide,
+    blockNumber: number,
+    limitPools?: string[],
+  ): Promise<null | ExchangePrices<BebopData>> {
+    this.tokensMap = (await this.getCachedTokens()) || {};
+
+    try {
+      const pools =
+        limitPools ??
+        (await this.getPoolIdentifiers(srcToken, destToken, side, blockNumber));
```

Comment:
Here safer to filter over limitPools if available, e.g. `pools = limitPools.filter(p => <pool is available>)`

Reply:
done